include/openenclave/internal/calls.h

```diff
@@ -473,7 +480,13 @@ oe_result_t oe_ecall(
  * @retval OE_UNEXPECTED An unexpected error occurred.
  *
  */
-oe_result_t oe_ocall(uint16_t func, uint64_t arg_in, uint64_t* arg_out);
+oe_result_t oe_ocall(
```

Comment:
What is the rationale for this new signature?

Reply:
The SDK as is currently written passes OCALL parameters for built-in functions (e.g., ) in pointers via host memory.

Take `oe_host_malloc` as an example. This function performs an OCALL whose `IN` parameter represents the memory size to allocate, and an `OUT` parameter that is the host-side pointer. In SGX, passing these parameters does not require any specialized marshaling (i.e., a parameters structure), the host can read and write the parameters via registers.

Consider now `oe_host_realloc`. The OCALL this function performs takes in a pointer to structure as its `IN` parameter and its `OUT` parameter is a pointer. In SGX, the `IN` pointer refers to a location in host memory (previously allocated), and is passed by value in a register. The `OUT` parameter pointer works in the same way.

So the first invocation to `oe_ocall` passes a value to the host (`IN`), and a pointer to the enclave (`OUT`). The second invocation passes a pointer as a value to the host (`IN`), and a pointer to the enclave (`OUT`).

This discrepancy in semantics is irrelevant in SGX because the enclave can access host memory and the OE SDK has direct control over the registers used for parameter passing.

In TrustZone, those assumptions don't hold. Both values and pointers must be marshaled, wrapped in proper descriptor structures, much like `oeedger8r` does. Passing a value (the size in `oe_malloc`) means the SDK must know to tell the TEE API that the value in question is a value. If it is a pointer, passing the pointer as a value won't work, because the pointer points to enclave memory, and allocating and filling a host buffer ahead of time, like in the case of `oe_host_realloc` also doesn't work because the enclave cannot access host memory. So the pointer cannot be passed as is, and the SDK must know that it's a pointer to something of a known size, then marshal the entire structure.

Reply:
Thanks for the explanation. I now see why we need a signature similar to 
https://github.com/openenclave/openenclave/blob/87397fdb3948b3a60e2195cf9f28fa1fad8bc111/include/openenclave/edger8r/enclave.h#L68-L74

However is seems that the boolean parameter is redundant. The function-enum (OE_OCALL_MALLOC, OE_OCALL_REALLOC) itself indicates whether to interpret the value as a pointer or not.

Note @mikbras is working on moving most of the core hand-written functions over to EDL.

Reply:
On the redundancy question, I agree; however, `_handle_call_builtin_function` would have to deduce that information for all hand-written `OE_OCALL_*`. I'd rather the caller tell `oe_ocall` if the `IN` value is a value or a pointer, instead of encoding that information effectively twice.

Perhaps I'm missing something and you had a different idea, though.

As to your note, that'd be nice!

Reply:
Thinking more about this, I'm leaning toward leaving this as-is for now, and waiting to see what Mike comes up with, then modifying this as required.

Thoughts?

Reply:
It would be good to keep oe_ecall and oe_ocall consistent.
Currently oe_ecall has the same signature as oe_ocall; but with an additional enclave parameter to indicate which enclave to invoke the method on:
https://github.com/openenclave/openenclave/blob/3dc3c9ff9e3336912cc88237f220f0326c6ce2e5/include/openenclave/internal/calls.h#L431-L435

Reply:
This is in fact a bug; the way I implemented `oe_ecall` only works for `oeedger8r` ECALL functions; all the others won't.

I'll fix it.

Reply:
Done.